### PR TITLE
[Fluidstack] Improve fluidstack provisioner for corner cases

### LINKED ISSUE #3259: [Fluidstack] Launching A100 machine fails due to the reboot after nvidia driver installation takes too long
`sky launch -c test-fluid --cloud fluidstack --gpus A100-80GB nvidia-smi`

https://gist.github.com/Michaelvll/a9fc4a7e335bcc293e302be60ca5142e

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes #3259.

It also fixes the issue when there are multiple existing instance with the name `*-head`, we now reuse those instances by renaming them.

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
  - [x] `sky launch --cloud fluidstack --gpus A100-80GB nvidia-smi -c test-fluid`
  - [x] `sky launch -c test-a100 --cloud fluidstack --gpus A100-80GB nvidia-smi --num-nodes 2`
  - [x] `sky launch -c test-a100 --cloud fluidstack --gpus A100-80GB:8 nvidia-smi`
  - [x] `sky launch -c test-h100 --cloud fluidstack --gpus H100 nvidia-smi`
  - [x] `sky launch -c test-h100 --cloud fluidstack --gpus H100:2 nvidia-smi`
  - [ ] `sky launch -c test-h100 --cloud fluidstack --gpus H100:8 nvidia-smi`
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/provision/fluidstack/instance.py [hunk 1]
```diff
@@ -89,18 +89,45 @@ def run_instances(region: str, cluster_name_on_cloud: str,
...
+            utils.FluidstackClient().rename(instance_id, new_name)
+        except Exception as e:
+            logger.warning(f'run_instances error: {e}')
+            raise
+
+    for inst_id, inst in exist_instances.items():
+        if head_instance_id is None:
+            head_instance_id = inst_id
+            rename(inst_id, f'{cluster_name_on_cloud}-head')
+            logger.info(f'Rename head node {head_instance_id}.')
```
Comment (cblmemo) [inline sky/provision/fluidstack/instance.py:115]:
```suggestion
            logger.info(f'Rename head node {head_instance_id} to {xxx}.')
```

File: sky/provision/fluidstack/instance.py [hunk 2]
```diff
@@ -89,18 +89,45 @@ def run_instances(region: str, cluster_name_on_cloud: str,
...
+        if inst_id != head_instance_id and inst['hostname'].endswith('-head'):
```
Comment (cblmemo) [inline sky/provision/fluidstack/instance.py:116]: Is this for the case when launching on an existing cluster where some node has been set to head node? in this case, should we keep the original head node?

File: sky/provision/fluidstack/fluidstack_utils.py [hunk 3]
```diff
@@ -239,3 +241,12 @@ def add_tags(self, instance_id: str, tags: Dict[str, str]):
         )
         raise_fluidstack_error(response)
         return response.json()
+
+    def rename(self, instance_id: str, hostname: str):
```
Comment (concretevitamin) [inline sky/provision/fluidstack/fluidstack_utils.py:245]: nit: return type

File: sky/provision/fluidstack/instance.py [hunk 4]
```diff
@@ -89,18 +89,47 @@ def run_instances(region: str, cluster_name_on_cloud: str,
...
+                f'{len(instances)} nodes, but {config.count} are '
+                'required. Please try terminate the cluster and retry.')
         if not instances:
             break
         logger.info(f'Waiting for {len(instances)} instances to be ready.')
         time.sleep(POLL_INTERVAL)
     exist_instances = _filter_instances(cluster_name_on_cloud, ['running'])
     head_instance_id = _get_head_instance_id(exist_instances)
 
+    def rename(instance_id, new_name):
```
Comment (concretevitamin) [inline sky/provision/fluidstack/instance.py:104]: nit: typing

File: sky/provision/fluidstack/instance.py [hunk 5]
```diff
@@ -89,18 +89,47 @@ def run_instances(region: str, cluster_name_on_cloud: str,
...
+        try:
+            utils.FluidstackClient().rename(instance_id, new_name)
+        except Exception as e:
+            logger.warning(f'run_instances error: {e}')
+            raise
+
+    for inst_id, inst in exist_instances.items():
```
Comment (concretevitamin) [inline sky/provision/fluidstack/instance.py:111]: (styleguide) Spell out: instance_id, instance, existing_instances

File: sky/provision/fluidstack/instance.py [hunk 6]
```diff
@@ -89,18 +89,47 @@ def run_instances(region: str, cluster_name_on_cloud: str,
...
+        if head_instance_id is None:
```
Comment (concretevitamin) [inline sky/provision/fluidstack/instance.py:112]: leave a comment: when would it be the case that `exist_instances` does not contain any instance names with `-head` suffix?

File: sky/provision/fluidstack/instance.py [hunk 7]
```diff
@@ -89,18 +89,47 @@ def run_instances(region: str, cluster_name_on_cloud: str,
...
+            head_instance_id = inst_id
+            instance_name = f'{cluster_name_on_cloud}-head'
+            rename(inst_id, instance_name)
+            logger.info(f'Rename head node {head_instance_id} to '
+                        f'{instance_name}')
+        if inst_id != head_instance_id and inst['hostname'].endswith('-head'):
+            logger.info(f'Rename worker node {inst_id}.')
```
Comment (concretevitamin) [inline sky/provision/fluidstack/instance.py:119]: nit: Be consistent in terms of logging + op ordering. "Renaming" here? L115-116 switch order?

File: sky/provision/fluidstack/instance.py (hunk 2 above)
Comment (concretevitamin) [reply sky/provision/fluidstack/instance.py:118]: Why would there be multiple `-head`? A bug on the CSP side? Worth a comment?

File: sky/provision/fluidstack/instance.py (hunk 6 above)
Comment (Michaelvll) [reply sky/provision/fluidstack/instance.py:112]: Added. Thanks!

File: sky/provision/fluidstack/instance.py [hunk 8]
```diff
@@ -101,25 +101,33 @@ def run_instances(region: str, cluster_name_on_cloud: str,
     exist_instances = _filter_instances(cluster_name_on_cloud, ['running'])
     head_instance_id = _get_head_instance_id(exist_instances)
 
-    def rename(instance_id, new_name):
+    def rename(instance_id: str, new_name: str) -> None:
```
Comment (concretevitamin) [inline sky/provision/fluidstack/instance.py:104]: Hmm, I think it's a bit weird we have to do it for Fluidstack but not other clouds. Renaming seems to assume reusing existing VMs is ok. Is it right to say an alternative is to down those old instances and relaunch everything? If so, I think it's fine to leave a NOTE/TODO and unblock the user first.

**Review by concretevitamin — APPROVED** [review]

**Comment by Michaelvll** [comment]
Just updated the instance creation logic, so that we don't wait infinitely for a instance directly being converted to terminated state from a pending state. PTAL @concretevitamin @cblmemo

**Comment by Michaelvll** [comment]
Merging this now to unblock our user : )
